akka-stream/src/main/scala/akka/stream/impl/SynchronousFilePublisher.scala

```diff
@@ -0,0 +1,124 @@
+private[akka] object SynchronousFilePublisher {
+  def props(f: File, completionPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int) = {
+    require(chunkSize > 0, s"chunkSize must be > 0 (was $chunkSize)")
+    require(initialBuffer > 0, s"initialBuffer must be > 0 (was $initialBuffer)")
+    require(maxBuffer >= initialBuffer, s"maxBuffer must be >= initialBuffer (was $maxBuffer)")
+
+    Props(classOf[SynchronousFilePublisher], f, completionPromise, chunkSize, initialBuffer, maxBuffer)
+  }
+
+  private final case object Continue extends DeadLetterSuppression
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFilePublisher(f: File, bytesReadPromise: Promise[Long], chunkSize: Int, initialBuffer: Int, maxBuffer: Int)
+  extends akka.stream.actor.ActorPublisher[ByteString]
+  with ActorLogging {
+
+  import SynchronousFilePublisher._
+
+  val buffs = new DirectByteBufferPool(chunkSize, maxBuffer)
+  var eofReachedAtOffset = Long.MinValue
+
+  var readBytesTotal = 0L
+  var availableChunks: Vector[ByteString] = Vector.empty
+
+  private var raf: RandomAccessFile = _
+  private var chan: FileChannel = _
+
+  override def preStart() = {
+    try {
+      raf = new RandomAccessFile(f, "r") // best way to express this in JDK6, OpenOption are available since JDK7
+      chan = raf.getChannel
+      readAndSignal(initialBuffer)
+    } catch {
+      case ex: Exception ⇒
+        onError(ex)
+    }
+
+    super.preStart()
+  }
+
+  def receive = {
+    case ActorPublisherMessage.Request(elements) ⇒ readAndSignal(maxBuffer)
+    case Continue                                ⇒ if (totalDemand > 0) readAndSignal(maxBuffer)
+    case ActorPublisherMessage.Cancel            ⇒ context.stop(self)
+  }
+
+  def readAndSignal(readAhead: Int): Unit =
+    if (isActive) {
+      // signal from available buffer right away
+      signalOnNexts()
+
+      // read chunks until readAhead is fulfilled
+      while (availableChunks.length < readAhead && !eofEncountered && isActive)
+        loadChunk()
+
+      self ! Continue
+    }
+
+  private def signalOnNexts(): Unit =
+    if (availableChunks.nonEmpty && totalDemand > 0) {
+      val ready = availableChunks.head
+      availableChunks = availableChunks.tail
+
+      onNext(ready)
+
+      if (totalDemand > 0) signalOnNexts()
+    } else if (eofEncountered) {
+      onComplete()
+    }
+
+  /** BLOCKING I/O READ */
+  def loadChunk() = try {
+    val buf = buffs.acquire()
+    buf.flip().clear()
+
+    // blocking read
+    val readBytes = chan.read(buf)
+
+    readBytes match {
+      case -1 | 0 ⇒
+        // had nothing to read into this chunk, will complete now
+        eofReachedAtOffset = chan.position
+        log.debug("No more bytes available to read (got `-1` from `read`), marking final bytes of file @ " + eofReachedAtOffset)
+
+      case _ ⇒
+        buf.flip()
+        readBytesTotal += readBytes
+        availableChunks :+= ByteString(buf).take(readBytes)
```

Comment:
Here it is, true. It remained from different impls. When moved to reading to byteArrays from InputStream it's needed.